### DISCUSSION TITLE
Widget with multiple habits

### DISCUSSION THREAD

**Discussion by sojusnik** [iSoron/uhabits#45] [Feature Requests] [answered]
A widget that contains all habits as a list would be very handy to tick habits off, like it's already implemented in [HabitBull](https://play.google.com/store/apps/details?id=com.oristats.habitbull) and shown [here](http://i.imgur.com/CQGFOIn.png).

**Comment by flowgunso** [comment]
The priority on this feature should be raised, it improves usability.

**Comment by mr-flibble** [comment]
I also think that this is an important feature that the application is missing.

**Comment by ketanarlulkar** [comment]
This will be a great feature, agree with @flowgunso, priority should be raised.

**Comment by vyu1** [comment]
Hi all, I haven't contributed to this project before, but is it okay if I try to work on doing this for specifically the Checkmark widget? I've found this feature lacking since first installing the app.

**Comment by iSoron** [comment]
Hi @vyu1, pull requests are very welcome. How are you planning to make the checkmark widget support multiple habits? Widgets are very limited on Android, but one thing we could use is the [StackView](http://www.broculos.net/2011/12/android-101-how-to-create-stackview.html) layout. If you have any other ideas, let me know.

**Comment by vyu1** [comment]
@iSoron I was thinking of making a new ListView layout (widgets don't support `RecyclerViews`) that looks similar to the image @sojusnik posted above, http://i.imgur.com/CQGFOIn.png and to `HabitsActivity`.
However, using a `StackView` could also be interesting, and I'm open to anyone's input on which I should try to implement.

**Comment by woj-tek** [comment]
+1

A widget with a view similar to main app, with possibility to select which habits to display (this would go hand in hand with #325)

**Comment by vyu1** [comment]
@woj-tek For the Checkmark widget, I think we're thinking of using a StackView so it wouldn't look like the main app.

**Comment by iSoron** [comment]
Thanks to @vyu1, it is now possible to stack multiple widgets together. This works not only for the checkmark widget, but for any other widget we currently have available. Please see the screenshot below.

At this time, I am not planning to add a widget similar to the app's main screen. This is mainly due to Android limitations. I already said this before, but Android widgets are truly limited on what they can do, or even in what they can look like. We already employ a number of hacks to work around these limitations and to make our current widgets look and behave as well as they do. I don't think it would be possible to recreate the main app screen on widgets. Even if we somehow managed to do it, I would expect it to be very sluggish, and have poor usability. That being said, anyone is free to try, of course, and I would be glad to review such a patch.

![screenshot_1529197063](https://user-images.githubusercontent.com/4699657/41503680-89cb5dc0-719f-11e8-95d6-c5a6ac480caa.png)

**Reply by j-dexx** [reply]
Hi,

How do you stack the widgets like that?  When I add a widget I can only select one habit, and if I try and add them on top of one another I can't as the other widget moves.

**Reply by iSoron** [reply]
@j-dexx This will become available in Loop 2.0.

**Reply by j-dexx** [reply]
Thanks @iSoron

**Reply by OliverStoll** [reply]
I understand the issues with the original question for ticking habits off, but what about a basic 'view' for your habits on the homescreen?

I have been searching for years for a habit tracking solution that both offers a concise widget to view your last 7 days of habits (such as HabitBull or HabitHub) and an API, and feel like this already great app would benefit immensly from this feature, even if it was not perfect or even close to perfect.

The posibility alone to view your habit progress and create a direct form of accountability on your homescreen is already worth a lot!

**Reply by mahibi** [reply]
> At this time, I am not planning to add a widget similar to the app's main screen. This is mainly due to Android limitations.

Is this still the same situation with newer Android versions?

**Reply by saeedesmaili** [reply]
Agreed with @OliverStoll . I think you should create a new feature request for this, and we can comment and upvote

**Reply by saeedesmaili** [reply]
Ticktick has this type of widget, but only for its paid subscribers. I hope we can get this on this app instead.
![Screenshot_20250322-082505](https://github.com/user-attachments/assets/57051462-0381-49e4-b7b3-5f565c8fd720)

**Comment by iSoron** [comment, marked as answer]
This feature has been implemented and will become available in Loop 2.0.

**Reply by drawtheliner** [reply]
> This feature has been implemented and will become available in Loop 2.0.

Good to hear!

Out of curiosity, which option was implemented: having the habits as a tick-able list (as in OP image), or as a stack?

Wouldn't a grid be better than a list? In other words, the equivalent of manually adding several Checkmark widgets.

About the stack idea, I'm probably missing something, but how is it useful, if it would only give access to one habit at a time?

**Comment by VA-MMXXII** [comment]
FWIW, with Google Pixel 5 and Android 11 I am able to select the "stacked" widgets, but they are impossible to access and simply don't work for me. 

Instead I have manually assembled grids of logically grouped Loop checkmark widgets. It's very difficult to get the alignment to look right, but you can do it.

So, for the checkmark widgets, perhaps what's needed instead of the current stacked concept is a table or grid of widgets that would be automatically aligned properly?

**Reply by drawtheliner** [reply]
The stacked widgets are interesting (swipe to see next, tap to select), but not practical for this use case in my opinion. Takes forever to find the desired habit.

I agree with @VA-MMXXII it would be very useful to have a widget that shows multiple checkmarks (for the selected habits). Note it wouldn't be the same view as in the app, as @iSoron mentioned earlier - in the app you get the full history, whereas in this multi-widget we'd only be able to interact with the current day.

This multi-widget could look similar to what it looks like right now to manually make a grid of checkbox widgets, difference being it would be much more practical and easy to add (and move/arrange).

Even better if this muti-widget had smart resize (changing proportions would reflow the checkboxes within the rectangular area) and option to add a group as per [this request/mockup](https://github.com/iSoron/uhabits/discussions/495#discussioncomment-1511545) by @530f6caa  (with the group name showing on top).

By the way, should this request be unmarked as solved? Unless most users are happy with current implementation... Or should I create a new request?
